### PR TITLE
SCMOD-5524: Add TaskMessage encoding variable

### PR DESCRIPTION
**Pull request by xbreizh**
https://portal.digitalsafe.net/browse/SCMOD-5524

### CODE REVIEW

**Comment by xbreizh** [comment]
Hi @dermot-hardy , there are 3 files that I copied and modified from the worker-framework:
QueueManager, ResultHandler, SimpleQueueConsumer

The intent is to modify those files directly in the worker-framework, but as this may have consequences on other services / workers, I thought that I would first copy them over to make it work then check with you if that was the right way to go.

From Simran's [draft pr](https://github.com/JobService/job-service/pull/256), I see that he also intended to made some changes in the jobtracking as well, should we go down that path also? If we do that, then we may have to update the [WorkerTaskData](https://github.com/WorkerFramework/worker-framework/blob/a6ca5d0e4f2a4e513449eb81f1f5e813142e3aac/worker-api/src/main/java/com/hpe/caf/api/worker/WorkerTaskData.java#L41). Can you please check and let me know if that makes sense please? Thx!

File: job-service/src/main/java/com/hpe/caf/services/job/queue/QueueServices.java [hunk 1]
```diff
@@ -88,14 +90,6 @@ public void sendMessage(
             } else {
                 throw new RuntimeException("Unknown taskDataEncoding");
             }
-        } else if (taskDataObj instanceof Map<?, ?>) {
```
Comment (dermot-hardy) [inline job-service/src/main/java/com/hpe/caf/services/job/queue/QueueServices.java:91]: I don't get why this is being dropped or how that squares with the code above.

File: job-service/src/main/java/com/hpe/caf/services/job/queue/QueueServices.java [hunk 2]
```diff
@@ -117,7 +111,7 @@ public void sendMessage(
                 workerAction.getTaskApiVersion(),
                 taskData,
                 TaskStatus.NEW_TASK,
-                Collections.<String, byte[]>emptyMap(),
+                Collections.emptyMap(),
```
Comment (dermot-hardy) [inline job-service/src/main/java/com/hpe/caf/services/job/queue/QueueServices.java]: Why is this changing?

File: worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java [hunk 3]
```diff
@@ -318,9 +317,9 @@ public void determineForwardingAction(TaskMessage proxiedTaskMessage, TaskInform
                 f.setFailureTime(new Date());
                 f.setFailureSource(getWorkerName(proxiedTaskMessage));
 
-                final byte[] taskData = proxiedTaskMessage.getTaskData();
+                final Object taskData = proxiedTaskMessage.getTaskData();
                 if (taskData != null) {
-                    f.setFailureMessage(new String(taskData, StandardCharsets.UTF_8));
```
Comment (dermot-hardy) [inline worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java]: `StandardCharsets` import not removed.

File: worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java [hunk 4]
```diff
@@ -136,14 +137,13 @@ private JobTrackingWorkerTask getJobTrackingWorkerTask(final WorkerTaskData work
     }
 
     private static byte[] validateVersionAndData(final WorkerTaskData workerTask, final int workerApiVersion)
-            throws InvalidTaskException, TaskRejectedException
+            throws TaskRejectedException, InvalidTaskException
```
Comment (dermot-hardy) [inline worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java]: Pointless change

File: worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java [hunk 5]
```diff
@@ -136,14 +137,13 @@ private JobTrackingWorkerTask getJobTrackingWorkerTask(final WorkerTaskData work
...
     {
         final int version = workerTask.getVersion();
         if (workerApiVersion < version) {
             throw new TaskRejectedException("Found task version " + version + ", which is newer than " +
                     workerApiVersion);
         }
-
```
Comment (dermot-hardy) [inline worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java]: Pointless change

File: worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java [hunk 6]
```diff
@@ -176,7 +176,6 @@ private JobTrackingWorkerTask getJobTrackingWorkerTask(final WorkerTaskData work
             } catch (final CodecException e) {
                 throw new InvalidTaskException("Invalid input message", e);
             }
-
```
Comment (dermot-hardy) [inline worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java]: Pointless change

File: worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java [hunk 7]
```diff
@@ -318,9 +317,9 @@ public void determineForwardingAction(TaskMessage proxiedTaskMessage, TaskInform
                 f.setFailureTime(new Date());
                 f.setFailureSource(getWorkerName(proxiedTaskMessage));
 
-                final byte[] taskData = proxiedTaskMessage.getTaskData();
+                final Object taskData = proxiedTaskMessage.getTaskData();
```
Comment (dermot-hardy) [inline worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java]: But `.getTaskData()` still returns a `byte[]`.  What is being achieved by this?

File: worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java [hunk 8]
```diff
@@ -318,9 +317,9 @@ public void determineForwardingAction(TaskMessage proxiedTaskMessage, TaskInform
...
                 if (taskData != null) {
-                    f.setFailureMessage(new String(taskData, StandardCharsets.UTF_8));
+                    f.setFailureMessage(new ObjectMapper().convertValue(taskData, String.class));
```
Comment (dermot-hardy) [inline worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java]: This shouldn't be changing.

File: worker-jobtracking/src/test/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactoryTest.java [hunk 9]
```diff
@@ -115,13 +123,14 @@ public void testTrackingReportTask() throws Exception
         final JobTrackingWorkerFactory workerFactory = createJobTrackingWorkerFactory(codec, reporter);
 
         //Test
+        workerFactory.getWorker(wtd).doWork();
```
Comment (dermot-hardy) [inline worker-jobtracking/src/test/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactoryTest.java]: I don't understand what we've achieving here by doing this twice.

File: job-service-scheduled-executor/src/main/java/com/hpe/caf/services/job/scheduled/executor/QueueServices.java [hunk 10]
```diff
@@ -67,44 +63,14 @@ public QueueServices(final Connection connection, final Channel publisherChannel
...
         final String taskId = UUID.randomUUID().toString();
 
-        //  Serialise the data payload. Encoding type is provided in the WorkerAction.
-        final byte[] taskData;
-
         //  Check whether taskData is in the form of a string or object, and serialise/decode as appropriate.
         LOG.debug("Validating the task data ...");
         final Object taskDataObj = workerAction.getTaskData();
-        
-        if (taskDataObj instanceof String) {
```
Comment (dermot-hardy) [inline job-service-scheduled-executor/src/main/java/com/hpe/caf/services/job/scheduled/executor/QueueServices.java:82]: You're using `taskDataObj` directly now but this code suggests that it could be a string, so that should be handled.  We should probably still go through a conversion, although of course it will be a different conversion than the one that used to be in this branch that was converting to a `byte[]`.

File: job-service-scheduled-executor/src/main/java/com/hpe/caf/services/job/scheduled/executor/QueueServices.java [hunk 11]
```diff
@@ -120,22 +86,24 @@ public void sendMessage(
...
+        LOG.debug("taskdata boko {}", taskDataObj);
+        final QueueTaskMessage queueTaskMessage = new QueueTaskMessage(
                 taskId,
                 workerAction.getTaskClassifier(),
                 workerAction.getTaskApiVersion(),
-                taskData,
+                taskDataObj,
                 TaskStatus.NEW_TASK,
-                Collections.<String, byte[]>emptyMap(),
+                Collections.emptyMap(),
```
Comment (dermot-hardy) [inline job-service-scheduled-executor/src/main/java/com/hpe/caf/services/job/scheduled/executor/QueueServices.java]: Is this change required?

File: job-service-scheduled-executor/src/main/java/com/hpe/caf/services/job/scheduled/executor/QueueServices.java [hunk 12]
```diff
@@ -120,22 +86,24 @@ public void sendMessage(
...
                 targetQueue,
-                trackingInfo);
+                trackingInfo
+        );
```
Comment (dermot-hardy) [inline job-service-scheduled-executor/src/main/java/com/hpe/caf/services/job/scheduled/executor/QueueServices.java]: Pointless change

File: job-service-scheduled-executor/src/main/java/com/hpe/caf/services/job/scheduled/executor/QueueServices.java [hunk 13]
```diff
@@ -120,22 +86,24 @@ public void sendMessage(
                 getStatusCheckIntervalMillis(ScheduledExecutorConfig.getStatusCheckIntervalSeconds()),
                 statusCheckUrl, ScheduledExecutorConfig.getTrackingPipe(), workerAction.getTargetPipe());
 
-        final TaskMessage taskMessage = new TaskMessage(
+        LOG.debug("taskdata boko {}", taskDataObj);
```
Comment (dermot-hardy) [inline job-service-scheduled-executor/src/main/java/com/hpe/caf/services/job/scheduled/executor/QueueServices.java]: Poor log message

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

File: job-service-scheduled-executor/src/main/java/com/hpe/caf/services/job/scheduled/executor/QueueServices.java (hunk 13 above)
Comment (xbreizh) [reply job-service-scheduled-executor/src/main/java/com/hpe/caf/services/job/scheduled/executor/QueueServices.java]: Indeed, just some testing logs. I removed it

File: job-service-scheduled-executor/src/main/java/com/hpe/caf/services/job/scheduled/executor/QueueServices.java (hunk 12 above)
Comment (xbreizh) [reply job-service-scheduled-executor/src/main/java/com/hpe/caf/services/job/scheduled/executor/QueueServices.java]: Restored

File: job-service-scheduled-executor/src/main/java/com/hpe/caf/services/job/scheduled/executor/QueueServices.java (hunk 11 above)
Comment (xbreizh) [reply job-service-scheduled-executor/src/main/java/com/hpe/caf/services/job/scheduled/executor/QueueServices.java]: I think that's a good change given that we don't need to specify the type and because the map is empty, but I understand that it's more some tidying up rather that being part of the current jira, so I restored

File: job-service-scheduled-executor/src/main/java/com/hpe/caf/services/job/scheduled/executor/QueueServices.java (hunk 10 above)
Comment (xbreizh) [reply job-service-scheduled-executor/src/main/java/com/hpe/caf/services/job/scheduled/executor/QueueServices.java:82]: Done, note that I extracted the method for better readability, let me know if you want me to restore.

File: worker-jobtracking/src/test/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactoryTest.java (hunk 9 above)
Comment (xbreizh) [reply worker-jobtracking/src/test/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactoryTest.java]: Restored

File: worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java (hunk 8 above)
Comment (xbreizh) [reply worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java]: Restored

File: worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java (hunk 7 above)
Comment (xbreizh) [reply worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java]: Restored

File: worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java (hunk 6 above)
Comment (xbreizh) [reply worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java]: Restored

File: worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java (hunk 5 above)
Comment (xbreizh) [reply worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java]: Restored

File: worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java (hunk 4 above)
Comment (xbreizh) [reply worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java]: Restored

File: worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java (hunk 3 above)
Comment (xbreizh) [reply worker-jobtracking/src/main/java/com/hpe/caf/worker/jobtracking/JobTrackingWorkerFactory.java]: Removed

File: job-service/src/main/java/com/hpe/caf/services/job/queue/QueueServices.java (hunk 2 above)
Comment (xbreizh) [reply job-service/src/main/java/com/hpe/caf/services/job/queue/QueueServices.java]: Restored

File: job-service/src/main/java/com/hpe/caf/services/job/queue/QueueServices.java (hunk 1 above)
Comment (xbreizh) [reply job-service/src/main/java/com/hpe/caf/services/job/queue/QueueServices.java:91]: Restored

**Comment by dermot-hardy** [comment]
Sorry I accidentally closed this.  Please raise a fresh PR for the branch.
